### ISSUE TITLE
knn-graph build time measurement - clarification

### ISSUE THREAD

**Issue by jimd6776** [facebookresearch/faiss#745]
Hi,
I wanted to ask for clarification on the "Construction of a knn-graph" benchmark in the following link. [https://github.com/facebookresearch/faiss/wiki/Low-level-benchmarks](https://github.com/facebookresearch/faiss/wiki/Low-level-benchmarks)
Why the build time for each knn-graph method changes in y axis? Also, the intersection measure in x axis does not vary according to nprobe? I can't understand exactly how the graph build time measurement is done here.

Thanks in advance.

**Comment by beauby** [comment]
Not sure I understand the question:
- "changes in y axis": do you mean between indexes, or for a given index?
- "the intersection measure in x axis does not vary according to nprobe": it does, the points of a given curve correspond to different values of nprobe for a given index.

**Comment by jimd6776** [comment]
- I mean for a given index. Because i thought that for a given number of cells (i.e. IVF16384) the build time does not vary. Why it does?
- If the points of a given curve vary according to nprobe as you say then choosing smaller nlist (i.e. IVF4096) would not give better accuracy results for the same nprobe number?

**Comment by beauby** [comment]
@jimd6776 
> I mean for a given index. Because i thought that for a given number of cells (i.e. IVF16384) the build time does not vary. Why it does?

- Build time here does not mean the build time of the index, but the build time of the k-nn graph, i.e. there is a step searching for all vectors their `k` nearest neighbors using the index, so the number of probes will change the total time (less probes → faster search for constant number of cells).

> If the points of a given curve vary according to nprobe as you say then choosing smaller nlist (i.e. IVF4096) would not give better accuracy results for the same nprobe number?

- Indeed. What makes you say it is not the case? The curves here all use the same nlist.

**Comment by jimd6776** [comment]
- I thought that nprobes variable only had to do with query search. For instance, in the following example the `index.train` is the index build while the `index.add` is the knn-graph build? Or something else?
[https://github.com/facebookresearch/faiss/blob/223d4e32d8dcc08342da6ca686f09edc00daa37d/tutorial/cpp/2-IVFFlat.cpp](https://github.com/facebookresearch/faiss/blob/223d4e32d8dcc08342da6ca686f09edc00daa37d/tutorial/cpp/2-IVFFlat.cpp)
- Basically i wanted to know how to measure the quality of a graph (or index i dont know) when comparing IVFFlat using different nlist values.

**Comment by beauby** [comment]
> I thought that nprobes variable only had to do with query search.

That is correct.

> For instance, in the following example the index.train is the index build while the index.add is the knn-graph build? Or something else?

No, in that example, no knn graph is built. The knn graph of a set of points is the graph the vertices of which are the points, and where there is an edge between vertices `u` and `v` iff `u` is within the k nearest neighbors (distance-wise) of `v` (or vice versa).

An index solves a more general problem: finding the k nearest neighbors of a set `Q` of points (the query), within a set `B` of points (the "content" of the index). Building a k-NN graph is done by having `B = Q`.

> Basically i wanted to know how to measure the quality of a graph (or index i dont know) when comparing IVFFlat using different nlist values.

You'll probably want to measure the recall@r for different values of nlist, for which you'll need an exact ground truth.

**Comment by jimd6776** [comment]
- So these commands have to do only with the Index build? Correct me if i am wrong; the "train" builds the clusters for the given nlist using train data, it's basically a kmeans clustering. What is the purpose of the "add"?

**Comment by beauby** [comment]
> So these commands have to do only with the Index build?

If by "these commands" you mean the example [cpp/2-IVFFlat.cpp](https://github.com/facebookresearch/faiss/blob/223d4e32d8dcc08342da6ca686f09edc00daa37d/tutorial/cpp/2-IVFFlat.cpp), then it is an example of building an index, training the quantizer, adding vectors to the index, and finally searching the index.

> Correct me if i am wrong; the "train" builds the clusters for the given nlist using train data, it's basically a kmeans clustering. 

That is correct (in the case of an IVF index).

> What is the purpose of the "add"?

The `add` method simply add vectors to the index, so that you can later search them.

**Comment by jimd6776** [comment]
Thank you very much.
